lib/Parse/ParseType.cpp

```diff
@@ -182,59 +183,46 @@ ParserResult<TypeRepr> Parser::parseType(Diag<> MessageID,
   if (ty.isNull())
     return nullptr;
 
-  // Parse a throws specifier.  'throw' is probably a typo for 'throws',
-  // but in local contexts we could just be at the end of a statement,
-  // so we need to check for the arrow.
-  ParserPosition beforeThrowsPos;
+  // Parse a throws specifier.
+  // Don't consume 'throws', if next token is not '->', so we can emit a more
+  // useful diagnostic when parsing a function decl.
   SourceLoc throwsLoc;
-  bool rethrows = false;
-  if (Tok.isAny(tok::kw_throws, tok::kw_rethrows) ||
-      (Tok.is(tok::kw_throw) && peekToken().is(tok::arrow))) {
-    if (Tok.is(tok::kw_throw)) {
+  if (Tok.isAny(tok::kw_throws, tok::kw_rethrows, tok::kw_throw) &&
+      peekToken().is(tok::arrow)) {
```

Comment:
This is definitely a behavior change for the recovery in a partial line:

```swift
let fn: () throws
```

Can you add tests for this case, and see if it's a net benefit?